src/modules/mtev_amqp.c

```diff
@@ -0,0 +1,566 @@
+      break;
+    }
+    case AMQP_CHANNEL_CLOSE_METHOD: {
+      amqp_channel_close_t *m = (amqp_channel_close_t *) x.reply.decoded;
+      mtevL(nlerr, "%s: server channel error %uh, message: %.*s\n",
+              context,
+              m->reply_code,
+              (int) m->reply_text.len, (char *) m->reply_text.bytes);
+      break;
+    }
+    default:
+      mtevL(nlerr, "%s: unknown server error, method id 0x%08X\n", context, x.reply.id);
+      break;
+    }
+    break;
+  }
+  return -1;
+}
+
+/* We use the fifo a bit unconventionally here, but we are
+   actually the only consumer on this, so there is no competitor
+   that could come in an confuse this.
+   We effectively peek (instead of pop) the message for sending
+   so that if it fails we can leak it enqueued where it is.
+   If we send without issue, then we dequeue and then free.
+ */
+static int
+drain_outbound_queue(struct amqp_conn *cc) {
+  int cnt = 0;
+  while(1) {
+    amqp_envelope_t *env = NULL, *sameenv = NULL;
+    struct ck_fifo_spsc_entry *entry;
+    int rv;
+    ck_fifo_spsc_dequeue_lock(&cc->outbound);
+    entry = CK_FIFO_SPSC_FIRST(&cc->outbound);
+    if(entry != NULL) env = entry->value;
+    ck_fifo_spsc_dequeue_unlock(&cc->outbound);
+    if(env) {
+      amqp_boolean_t mandatory = (uintptr_t)env & ENV_MANDATORY;
+      amqp_boolean_t immediate = (uintptr_t)env & ENV_IMMEDIATE;
+      env = (void *)((uintptr_t)env & ~(sizeof(uintptr_t)-1));
+      rv = amqp_basic_publish(cc->conn, 1, env->exchange, env->routing_key,
+                              mandatory, immediate,
+                              &env->message.properties, env->message.body);
+      if(rv != AMQP_STATUS_OK) {
+        mtevL(nlerr, "basic_publish failed: %d\n", rv);
+        return -1;
+      }
+    }
+    ck_fifo_spsc_dequeue_lock(&cc->outbound);
+    bool found = ck_fifo_spsc_dequeue(&cc->outbound, &sameenv);
+    ck_fifo_spsc_dequeue_unlock(&cc->outbound);
+    if(!found) break;
+    sameenv = (void *)((uintptr_t)sameenv & ~(sizeof(uintptr_t)-1));
+    mtevAssert(env == sameenv);
+    amqp_destroy_envelope(env);
+    free(env);
+    cnt++;
+  }
+  return cnt;
+}
+/* Because amqp_connection_state_t's are not thread safe, we must read and
+   write to this connection in the same thread... here.
+   This is problematic b/c knowing we have data to write and knowing we can
+   read data use different fascilitites and thus we have to spin.
+
+   So.. we will wait for "up to" 50ms to receive a message and then drain
+   our outbound queue.  If the inbound channel is idle, this means that we
+   will send outbound data in tight bunches every 50ms.
+
+   This function does work until it can't make progress (error) and then
+   returns to the caller (which will shutdown the connection and attempt to
+   bring it back online).
+ */
+static void
+handleconn(struct amqp_conn *conn) {
+  while(1) {
+    struct timeval small_sleep = { 0, 100000 }; /* 50ms */
```

Comment:
Well... the param to consume message isn't const, so I worry it might update it.  I always want it to be 1s, so it seemed safest to explicitly ensure that before each call.